### PR TITLE
Creation of a docker development environment

### PR DESCRIPTION
**Pull request by NotFish232**
A development environment for ion using docker compose.

### CODE REVIEW

**Comment by Laur04** [comment]
@shahsalonik @Rushilwiz @nkenschaft You'll be using this the most next year, so test it out when you get some time and let me know your thoughts. 

Also, @smayya337 @etnguyen03

File: config/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM ubuntu:latest
+RUN apt-get -y update && apt-get -y clean 
+RUN apt-get -y install sudo
+RUN apt-get -y install git
+RUN sudo apt-get -y install python3
```
Comment (Laur04) [inline config/Dockerfile:5]: You are running everything as root, so you don't need sudo.

**Review by Laur04 — CHANGES_REQUESTED** [review]
There needs to be a documentation page included in this PR.

File: config/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+RUN sudo apt-get -y install python3-pip
+RUN sudo apt-get -y install python-is-python3
+RUN sudo apt-get -y install libpq-dev 
+RUN sudo apt-get -y install ruby-full rubygems
+RUN gem install sass
```
Comment (etnguyen03) [inline config/Dockerfile:10]: Every layer you add is an extra layer to the Docker image.

You want to string together as much as you can with `&&`.. i.e.

```suggestion
RUN apt-get -y update && apt-get -y clean && apt-get -y install sudo git python3 python3-pip python-is-python3 libpq-dev ruby-full rubygems && gem install sass
```

File: config/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    depends_on:
+    - redis
+    - postgres
+networks:
+  intranet_net:
+      driver: bridge
+volumes:
+  pgdata:
+  <DEFAULT_VOLUME_NAME>:
+    external: true
```
Comment (Laur04) [inline config/docker-compose.yaml:44]:
```suggestion
```

File: config/docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    image: application
+    networks:
+    - intranet_net
+    ports:
+    - 8080:8080
+    build: 
+      context: ../
+      dockerfile: ./config/Dockerfile
+    volumes: 
+    - <DEFAULT_VOLUME_NAME>:/ion
```
Comment (Laur04) [inline config/docker-compose.yaml:34]:
```suggestion
    - <PATH_TO_ION_REPO>:/ion
```

File: config/docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+version: '3.9'
```
Comment (Laur04) [inline config/docker-compose.yaml:1]:
```suggestion
version: '3.9'

```

File: config/docker-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+services:
+  redis:
+    container_name: server
+    image: redis:latest
+    networks:
+    - intranet_net
+    ports:
+    - 6379:6379
```
Comment (Laur04) [inline config/docker-compose.yaml:9]:
```suggestion
    expose:
    - 6379
    
```

File: config/docker-compose.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  postgres:
+    container_name: database
+    image: postgres:latest
+    networks:
+    - intranet_net
+    ports:
+    - 5432:5432
```
Comment (Laur04) [inline config/docker-compose.yaml:16]:
```suggestion
    expose:
    - 5432
```

File: config/docker-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    environment:
+    - POSTGRES_DB=ion
+    - POSTGRES_USER=ion
+    - POSTGRES_PASSWORD=pwd
+    volumes:
+    - pgdata:/var/lib/postgresql/data
```
Comment (Laur04) [inline config/docker-compose.yaml:22]:
```suggestion
    - pgdata:/var/lib/postgresql/data
    
```

File: config/docker-compose.yaml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    ports:
+    - 8080:8080
+    build: 
+      context: ../
+      dockerfile: ./config/Dockerfile
+    volumes: 
+    - <DEFAULT_VOLUME_NAME>:/ion
+    depends_on:
+    - redis
+    - postgres
```
Comment (Laur04) [inline config/docker-compose.yaml:37]:
```suggestion
    - postgres
    
```

File: config/docker-compose.yaml [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+networks:
+  intranet_net:
+      driver: bridge
```
Comment (Laur04) [inline config/docker-compose.yaml:40]:
```suggestion
      driver: bridge
      
```

File: config/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,13 @@
+FROM ubuntu:latest
+RUN apt-get -y update && apt-get -y clean && apt-get -y install sudo git python3 python3-pip python-is-python3 libpq-dev ruby-full rubygems && gem install sass
+ENV TZ America/New_York
+RUN ln -s /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
+RUN apt-get -y install kinit
+COPY config/krb5.conf /etc/krb5.conf
+RUN DEBIAN_FRONTEND=noninteractive apt-get -y install krb5-user
+COPY requirements.txt .
+RUN pip install -r requirements.txt
+RUN rm requirements.txt
```
Comment (Laur04) [inline config/Dockerfile:10]:
```suggestion
RUN pip install -r requirements.txt && rm requirements.txt
```

File: config/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,13 @@
+FROM ubuntu:latest
+RUN apt-get -y update && apt-get -y clean && apt-get -y install sudo git python3 python3-pip python-is-python3 libpq-dev ruby-full rubygems && gem install sass
+ENV TZ America/New_York
+RUN ln -s /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
+RUN apt-get -y install kinit
+COPY config/krb5.conf /etc/krb5.conf
+RUN DEBIAN_FRONTEND=noninteractive apt-get -y install krb5-user
```
Comment (Laur04) [inline config/Dockerfile:7]:
```suggestion
COPY config/krb5.conf /etc/krb5.conf
RUN DEBIAN_FRONTEND=noninteractive apt-get -y install krb5-user kinit
```

File: config/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,13 @@
+FROM ubuntu:latest
+RUN apt-get -y update && apt-get -y clean && apt-get -y install sudo git python3 python3-pip python-is-python3 libpq-dev ruby-full rubygems && gem install sass
```
Comment (Laur04) [inline config/Dockerfile:2]:
```suggestion
RUN apt-get -y update && apt-get -y clean && apt-get -y install git python3 python3-pip python-is-python3 libpq-dev ruby-full rubygems && gem install sass
```

File: config/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,13 @@
+FROM ubuntu:latest
```
Comment (Laur04) [inline config/Dockerfile:1]:
```suggestion
FROM ubuntu:latest

```

File: config/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+RUN apt-get -y update && apt-get -y clean && apt-get -y install sudo git python3 python3-pip python-is-python3 libpq-dev ruby-full rubygems && gem install sass
+ENV TZ America/New_York
+RUN ln -s /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
+RUN apt-get -y install kinit
+COPY config/krb5.conf /etc/krb5.conf
+RUN DEBIAN_FRONTEND=noninteractive apt-get -y install krb5-user
+COPY requirements.txt .
+RUN pip install -r requirements.txt
+RUN rm requirements.txt
+EXPOSE 8080
```
Comment (Laur04) [inline config/Dockerfile:11]:
```suggestion
```

You already do this in `docker-compose`.

File: config/documentation.md [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+clone the ion git repository, found [here](https://github.com/tjcsl/ion), into the dev environment.
+Next, open the dev environment in vscode.
+run 'cd config' to change your directory into config
+In the config directory, edit the file 'make_admin.py', substituting '<YOURUSERNAME>' for your ion username
+Then, to find the name of your dev environment, go to docker desktop and to the tab 'DEV ENVIRONMENTS'
+the name should be in the form of 'ion-< >_< >'.
+Edit the file 'docker-compose.yaml', replacing both instances of <DEFAULT_VOLUME_NAME> to 
+'volume-(dev environment name)'
+Finally, run 'docker-compose up' and wait while the development environment is set up.
+The application will be running at localhost:8080.
```
Comment (Laur04) [inline config/documentation.md:14]:
```suggestion
# Create an Ion Development Enviornment

## Prerequisites
- [Docker](https://www.docker.com) is the containization framework that runs Ion's individual services. Installation is OS-specific and instructions can be found [here](ttps://www.docker.com/products/docker-desktop/). On Windows and Mac, install Docker Desktop.
- [GitHub](https://github.com) is the version control system used by the CSL. Make sure that you have an account and an SSH key tied to that account that will allow you to push and pull code.


## Set Up
1. Create your own fork of the [``tjcsl/ion`` repository](https://github.com/tjcsl/ion.git).
2. Clone the Ion repositiory from your Ion fork by running ``git clone git@github.com:<YOUR_GITHUB_USERNAME>/ion.git``.
3. Inside the Ion repository, edit the ``config/make_admin.py`` file, replacing "\<YOURUSERNAME\>" with your Ion username.
4. Inside the Ion repository, edit the ``config/docker-compose.yml`` file, replacing \<PATH_TO_ION_REPO\> with the path to where your Ion repository lives on your host machine.
5. Run ``cd config``.
5. Run ``docker-compose up -d``.


## Post Set Up
Navigate to http://localhost:8000 in the web browser of your choice. You might have to wait up to 60 seconds the first time that you create the container. When presented with the login page, log in with your Ion username and the password "nofish" (without the quotes).


## Useful Commands
### Interacting with the application:
If you need to run a Django command like ``makemigrations``, ``collectstatic`` or ``shell_plus``, run ``docker exec -it application bash`` in your terminal. That wlll give you a shell into the application container. You can also use this to run scripts like ``build_sources.sh``. If you need to view the output from or restart ``runserver``, run ``docker attach application``. 
```

**Review by Laur04 — CHANGES_REQUESTED** [review]

File: config/docker-compose.yaml (hunk 4 above)
Comment (smayya337) [reply config/docker-compose.yaml:34]: Even better than `<PATH_TO_ION_REPO>:/ion`: just use `../:/ion/:z`. (The `z` is admittedly more for me than anyone else; it keeps SELinux from restricting my access.)

File: config/documentation.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
+# Create an Ion Development Enviornment
+
+## Prerequisites
+- [Docker](https://www.docker.com) is the containization framework that runs Ion's individual services. Installation is OS-specific and instructions can be found [here](ttps://www.docker.com/products/docker-desktop/). On Windows and Mac, install Docker Desktop.
```
Comment (Laur04) [inline config/documentation.md:4]:
```suggestion
- [Docker](https://www.docker.com) is the containerization framework that runs Ion's individual services. Installation is OS-specific and instructions can be found [here](ttps://www.docker.com/products/docker-desktop/). On Windows and Mac, install Docker Desktop.
```

My bad on the spelling.

**Review by Laur04 — CHANGES_REQUESTED** [review]
Okay two very minor changes.

**Review by Laur04 — APPROVED** [review]

**Comment by Laur04** [comment]
Closing in favor of #1278

**Pull request closed without merging**